cloudinit/sources/DataSourceVMware.py

```diff
@@ -0,0 +1,780 @@
+        return handle_returned_guestinfo_val(key, stdout)
+    except ProcessExecutionError as error:
+        if error.stderr == NOVAL:
+            LOG.debug("No value found for key %s", key)
+        else:
+            util.logexc(
+                LOG,
+                "Failed to get guestinfo value for key %s: %s",
+                key,
+                error,
+            )
+    except Exception:
+        util.logexc(
+            LOG,
+            "Unexpected error while trying to get "
+            + "guestinfo value for key %s",
+            key,
+        )
+
+    return None
+
+
+def guestinfo_set_value(key, value, vmware_rpctool=VMWARE_RPCTOOL):
+    """
+    Sets a guestinfo value for the specified key. Set value to an empty string
+    to clear an existing guestinfo key.
+    """
+
+    # If value is an empty string then set it to a single space as it is not
+    # possible to set a guestinfo key to an empty string. Setting a guestinfo
+    # key to a single space is as close as it gets to clearing an existing
+    # guestinfo key.
+    if value == "":
+        value = " "
+
+    LOG.debug("Setting guestinfo key=%s to value=%s", key, value)
+
+    try:
+        subp([vmware_rpctool, ("info-set guestinfo.%s %s" % (key, value))])
+        return True
+    except ProcessExecutionError as error:
+        util.logexc(
+            LOG,
+            "Failed to set guestinfo key=%s to value=%s: %s",
+            key,
+            value,
+            error,
+        )
+    except Exception:
+        util.logexc(
+            LOG,
+            "Unexpected error while trying to set "
+            + "guestinfo key=%s to value=%s",
+            key,
+            value,
+        )
+
+    return None
+
+
+def guestinfo_clear_keys(keys, vmware_rpctool=VMWARE_RPCTOOL):
+    """
+    guestinfo_clear_keys clears guestinfo of all of the keys in the given list.
+    each key will have its value set to "---". Since the value is valid YAML,
+    cloud-init can still read it if it tries.
+    """
+    if not keys:
+        return
+    if not type(keys) in (list, tuple):
+        keys = [keys]
+    for key in keys:
+        LOG.info("clearing guestinfo.%s", key)
+        if not guestinfo_set_value(
+            key, GUESTINFO_EMPTY_YAML_VAL, vmware_rpctool
+        ):
+            LOG.error("failed to clear guestinfo.%s", key)
+        LOG.info("clearing guestinfo.%s.encoding", key)
+        if not guestinfo_set_value(key + ".encoding", "", vmware_rpctool):
+            LOG.error("failed to clear guestinfo.%s.encoding", key)
+
+
+def load(data):
+    """
+    load first attempts to unmarshal the provided data as JSON, and if
+    that fails then attempts to unmarshal the data as YAML. If data is
+    None then a new dictionary is returned.
+    """
+    if not data:
+        return {}
+    try:
+        return json.loads(data)
+    except (json.JSONDecodeError, TypeError):
+        return safeyaml.load(data)
```

Comment:
These could use both util.load_json and util.load_yaml, which would also allow handling binary data.... but maybe you don't want all the extra baked in decoding operations?
```suggestion
    try:
        return util.load_json(data)
    except (json.JSONDecodeError, TypeError):
        return util.load_yaml(data)
```

Reply:
I'm sure those are fine. I created a lot of this back in early 2018, so I'm not sure if those functions were available then. And if they were, it's likely I just did what I needed inside the DS to make it work without worrying too much about finding the helpers. Thanks for pointing this out.